jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeNode.kt

```diff
@@ -65,27 +76,69 @@ class S3TreeDirectoryNode(private val bucket: S3VirtualBucket, parent: S3TreeDir
             ?.filterNot { it.key() == key }
-            ?.map { S3TreeObjectNode(bucketName, this, it.key(), it.size(), it.lastModified()) as S3TreeNode }
+            ?.map { S3TreeObjectNode(bucket, this, it.key(), it.size(), it.lastModified()) as S3TreeNode }
             ?: emptyList()
 
         return (folders + s3Objects).sortedBy { it.key } + continuation
     }
-
-    fun removeAllChildren() {
-        cachedList = listOf()
-    }
 }
 
 private val fileTypeRegistry = FileTypeRegistry.getInstance()
 
-class S3TreeObjectNode(bucketName: String, parent: S3TreeDirectoryNode?, key: String, val size: Long, val lastModified: Instant) :
-    S3TreeNode(bucketName, parent, key) {
+open class S3TreeObjectNode(val bucket: S3VirtualBucket, parent: S3LazyLoadParentNode?, key: String, val size: Long, val lastModified: Instant) :
+    S3LazyLoadParentNode(bucket.name, parent, key) {
 
+    var showHistory: Boolean = false
+    var responseIterator: Iterator<ListObjectVersionsResponse>? = null
     private val fileType = fileTypeRegistry.getFileTypeByFileName(name)
 
     init {
         fileType.takeIf { it !is UnknownFileType }?.icon.let { icon = it }
     }
+
+    override fun loadObjects(continuationMarker: String?): List<S3TreeNode> {
+        if (showHistory) {
+            responseIterator = responseIterator ?: runBlocking {
+                bucket.listObjectVersionsPaginated(key)
+            }.iterator()
+
+            val nextPage = responseIterator
+                ?.next()
+                ?.versions()
+                ?.map { S3TreeObjectVersionNode(bucket, parent, key, it.size(), it.lastModified(), it.versionId()) as S3TreeNode }
+                ?: emptyList()
+
+            if (responseIterator?.hasNext() == true) {
+                return nextPage + S3TreeContinuationNode(
+                    bucketName,
+                    this,
+                    "${this.key}/${message("s3.load_more")}",
+                    (nextPage.last() as S3TreeObjectVersionNode).versionId
+                )
+            }
+            return nextPage
+        }
+        return emptyList()
+    }
+}
+
+class S3TreeObjectVersionNode(bucket: S3VirtualBucket, parent: S3LazyLoadParentNode?, key: String, size: Long, lastModified: Instant, val versionId: String) :
+    S3TreeObjectNode(bucket, parent, key, size, lastModified) {
+
+    override fun getName(): String {
+        // For not versioned buckets api can return versionId as literal 'null' so we avoid propagating null string to UI.
+        val versionIdPostfix = if (versionId != "null") versionId else ""
+        val name = FileUtilRt.getNameWithoutExtension(super.getName())
+        val originalExtension = FileUtilRt.getExtension(super.getName())
+        val extension = if (originalExtension.isNotBlank()) ".$originalExtension" else ""
+
+        return "${name}_$versionIdPostfix$extension"
+    }
+
+    private val fileType = fileTypeRegistry.getFileTypeByFileName(super.getName())
```

Comment:
If this class takes `S3TreeObjectNode` as a parent, call parent.getName()

Reply:
Where is this used? It is private and not overriding the super val.

Reply:
It wasn't used after all I removed it completely